// @storybook/addon-ondevice-controls/dist/components/ModalPortal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModalPortal = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const portal_1 = require("@gorhom/portal");
const react_native_1 = require("react-native");
const modalSupported = react_native_1.Platform.OS !== 'macos';
const ModalPortal = ({ children, visible, usePortal, ...props }) => {
    if (modalSupported && !usePortal) {
        return ((0, jsx_runtime_1.jsx)(react_native_1.Modal, { visible: visible, ...props, children: children }));
    }
    if (!visible)
        return null;
    return ((0, jsx_runtime_1.jsx)(portal_1.Portal, { hostName: "storybook-lite-ui-root", children: (0, jsx_runtime_1.jsx)(react_native_1.View, { style: {
                flex: 1,
                position: 'absolute',
                top: 0,
                left: 0,
                right: 0,
                bottom: 0,
                zIndex: 1000,
            }, children: children }) }));
};
exports.ModalPortal = ModalPortal;
